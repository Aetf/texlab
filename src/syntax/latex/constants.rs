pub const INCLUDE_COMMANDS: &'static [&'static str] = &[
    "\\include",
    "\\input",
    "\\bibliography",
    "\\addbibresource",
    "\\usepackage",
    "\\documentclass",
];

pub const LABEL_DEFINITION_COMMANDS: &'static [&'static str] = &["\\label"];

pub const LABEL_REFERENCE_COMMANDS: &'static [&'static str] = &["\\ref", "\\autoref", "\\eqref"];

pub const SECTION_COMMANDS: &'static [&'static str] = &[
    "\\chapter",
    "\\chapter*",
    "\\section",
    "\\section*",
    "\\subsection",
    "\\subsection*",
    "\\subsubsection",
    "\\subsubsection*",
    "\\paragraph",
    "\\paragraph*",
    "\\subparagraph",
    "\\subparagraph*",
];

pub const CITATION_COMMANDS: &'static [&'static str] = &[
    "\\cite",
    "\\cite*",
    "\\Cite",
    "\\nocite",
    "\\citet",
    "\\citep",
    "\\citet*",
    "\\citep*",
    "\\citeauthor",
    "\\citeauthor*",
    "\\Citeauthor",
    "\\Citeauthor*",
    "\\citetitle",
    "\\citetitle*",
    "\\citeyear",
    "\\citeyear*",
    "\\citedate",
    "\\citedate*",
    "\\citeurl",
    "\\fullcite",
    "\\citeyearpar",
    "\\citealt",
    "\\citealp",
    "\\citetext",
    "\\parencite",
    "\\parencite*",
    "\\Parencite",
    "\\footcite",
    "\\footfullcite",
    "\\footcitetext",
    "\\textcite",
    "\\Textcite",
    "\\smartcite",
    "\\Smartcite",
    "\\supercite",
    "\\autocite",
    "\\Autocite",
    "\\autocite*",
    "\\Autocite*",
    "\\volcite",
    "\\Volcite",
    "\\pvolcite",
    "\\Pvolcite",
    "\\fvolcite",
    "\\ftvolcite",
    "\\svolcite",
    "\\Svolcite",
    "\\tvolcite",
    "\\Tvolcite",
    "\\avolcite",
    "\\Avolcite",
    "\\notecite",
    "\\notecite",
    "\\pnotecite",
    "\\Pnotecite",
    "\\fnotecite",
];
