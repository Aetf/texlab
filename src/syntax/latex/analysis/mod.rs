mod command;
mod distro_file;
mod environment;
mod explicit_link;
mod implicit_link;
mod label_number;
mod theorem;
mod types;

use crate::syntax::latex;

pub use self::types::*;
use self::{
    command::analyze_command,
    environment::analyze_begin,
    explicit_link::{analyze_import, analyze_include},
    implicit_link::analyze_implicit_links,
    label_number::analyze_label_number,
    theorem::analyze_theorem_definition,
};

pub fn analyze(context: &mut LatexAnalyzerContext, root: &latex::SyntaxNode) {
    analyze_implicit_links(context);
    for node in root.descendants() {
        analyze_command(context, node)
            .or_else(|| analyze_begin(context, node))
            .or_else(|| analyze_include(context, node))
            .or_else(|| analyze_import(context, node))
            .or_else(|| analyze_label_number(context, node))
            .or_else(|| analyze_theorem_definition(context, node));
    }
}
