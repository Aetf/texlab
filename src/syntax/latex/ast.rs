use texlab_protocol::RangeExt;
use crate::syntax::text::{Span, SyntaxNode};
use itertools::Itertools;
use lsp_types::Range;
use std::sync::Arc;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum LatexTokenKind {
    Word,
    Command,
    Math,
    Comma,
    BeginGroup,
    EndGroup,
    BeginOptions,
    EndOptions,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexToken {
    pub span: Span,
    pub kind: LatexTokenKind,
}

impl LatexToken {
    pub fn new(span: Span, kind: LatexTokenKind) -> Self {
        Self { span, kind }
    }

    pub fn text(&self) -> &str {
        &self.span.text
    }
}

impl SyntaxNode for LatexToken {
    fn range(&self) -> Range {
        self.span.range()
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Default)]
pub struct LatexRoot {
    pub children: Vec<LatexContent>,
}

impl LatexRoot {
    pub fn new(children: Vec<LatexContent>) -> Self {
        Self { children }
    }
}

impl SyntaxNode for LatexRoot {
    fn range(&self) -> Range {
        if self.children.is_empty() {
            Range::new_simple(0, 0, 0, 0)
        } else {
            Range::new(
                self.children[0].start(),
                self.children[self.children.len() - 1].end(),
            )
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum LatexContent {
    Group(Arc<LatexGroup>),
    Command(Arc<LatexCommand>),
    Text(Arc<LatexText>),
    Comma(Arc<LatexComma>),
    Math(Arc<LatexMath>),
}

impl LatexContent {
    pub fn accept<T: LatexVisitor>(&self, visitor: &mut T) {
        match self {
            LatexContent::Group(group) => visitor.visit_group(Arc::clone(&group)),
            LatexContent::Command(command) => visitor.visit_command(Arc::clone(&command)),
            LatexContent::Text(text) => visitor.visit_text(Arc::clone(&text)),
            LatexContent::Comma(comma) => visitor.visit_comma(Arc::clone(&comma)),
            LatexContent::Math(math) => visitor.visit_math(Arc::clone(&math)),
        }
    }
}

impl SyntaxNode for LatexContent {
    fn range(&self) -> Range {
        match self {
            LatexContent::Group(group) => group.range(),
            LatexContent::Command(command) => command.range(),
            LatexContent::Text(text) => text.range(),
            LatexContent::Comma(comma) => comma.range(),
            LatexContent::Math(math) => math.range(),
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum LatexGroupKind {
    Group,
    Options,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexGroup {
    pub range: Range,
    pub left: LatexToken,
    pub children: Vec<LatexContent>,
    pub right: Option<LatexToken>,
    pub kind: LatexGroupKind,
}

impl LatexGroup {
    pub fn new(
        left: LatexToken,
        children: Vec<LatexContent>,
        right: Option<LatexToken>,
        kind: LatexGroupKind,
    ) -> Self {
        let end = if let Some(ref right) = right {
            right.end()
        } else if !children.is_empty() {
            children[children.len() - 1].end()
        } else {
            left.end()
        };

        Self {
            range: Range::new(left.start(), end),
            left,
            children,
            right,
            kind,
        }
    }
}

impl SyntaxNode for LatexGroup {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexCommand {
    pub range: Range,
    pub name: LatexToken,
    pub options: Vec<Arc<LatexGroup>>,
    pub args: Vec<Arc<LatexGroup>>,
    pub groups: Vec<Arc<LatexGroup>>,
}

impl LatexCommand {
    pub fn new(
        name: LatexToken,
        options: Vec<Arc<LatexGroup>>,
        args: Vec<Arc<LatexGroup>>,
    ) -> Self {
        let groups: Vec<Arc<LatexGroup>> = args
            .iter()
            .chain(options.iter())
            .sorted_by_key(|group| group.range.start)
            .map(Arc::clone)
            .collect();

        let end = if let Some(group) = groups.last() {
            group.end()
        } else {
            name.end()
        };

        Self {
            range: Range::new(name.start(), end),
            name,
            options,
            args,
            groups,
        }
    }

    pub fn short_name_range(&self) -> Range {
        Range::new_simple(
            self.name.start().line,
            self.name.start().character + 1,
            self.name.end().line,
            self.name.end().character,
        )
    }

    pub fn extract_text(&self, index: usize) -> Option<&LatexText> {
        if self.args.len() > index && self.args[index].children.len() == 1 {
            if let LatexContent::Text(ref text) = self.args[index].children[0] {
                Some(text)
            } else {
                None
            }
        } else {
            None
        }
    }

    pub fn extract_word(&self, index: usize) -> Option<&LatexToken> {
        let text = self.extract_text(index)?;
        if text.words.len() == 1 {
            Some(&text.words[0])
        } else {
            None
        }
    }

    pub fn has_word(&self, index: usize) -> bool {
        self.extract_word(index).is_some()
    }

    pub fn extract_comma_separated_words(&self, index: usize) -> Vec<&LatexToken> {
        let mut words = Vec::new();
        for child in &self.args[index].children {
            if let LatexContent::Text(text) = child {
                for word in &text.words {
                    words.push(word);
                }
            }
        }
        words
    }

    pub fn has_comma_separated_words(&self, index: usize) -> bool {
        if self.args.len() <= index {
            return false;
        }

        for node in &self.args[index].children {
            match node {
                LatexContent::Text(_) | LatexContent::Comma(_) => (),
                LatexContent::Command(_) | LatexContent::Group(_) | LatexContent::Math(_) => {
                    return false;
                }
            }
        }
        true
    }
}

impl SyntaxNode for LatexCommand {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexText {
    pub range: Range,
    pub words: Vec<LatexToken>,
}

impl LatexText {
    pub fn new(words: Vec<LatexToken>) -> Self {
        Self {
            range: Range::new(words[0].start(), words[words.len() - 1].end()),
            words,
        }
    }
}

impl SyntaxNode for LatexText {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexComma {
    pub token: LatexToken,
}

impl LatexComma {
    pub fn new(token: LatexToken) -> Self {
        Self { token }
    }
}

impl SyntaxNode for LatexComma {
    fn range(&self) -> Range {
        self.token.range()
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexMath {
    pub token: LatexToken,
}

impl LatexMath {
    pub fn new(token: LatexToken) -> Self {
        Self { token }
    }
}

impl SyntaxNode for LatexMath {
    fn range(&self) -> Range {
        self.token.range()
    }
}

pub trait LatexVisitor {
    fn visit_root(&mut self, root: Arc<LatexRoot>);

    fn visit_group(&mut self, group: Arc<LatexGroup>);

    fn visit_command(&mut self, command: Arc<LatexCommand>);

    fn visit_text(&mut self, text: Arc<LatexText>);

    fn visit_comma(&mut self, comma: Arc<LatexComma>);

    fn visit_math(&mut self, math: Arc<LatexMath>);
}

pub struct LatexWalker;

impl LatexWalker {
    pub fn walk_root<T: LatexVisitor>(visitor: &mut T, root: Arc<LatexRoot>) {
        for child in &root.children {
            child.accept(visitor);
        }
    }

    pub fn walk_group<T: LatexVisitor>(visitor: &mut T, group: Arc<LatexGroup>) {
        for child in &group.children {
            child.accept(visitor);
        }
    }

    pub fn walk_command<T: LatexVisitor>(visitor: &mut T, command: Arc<LatexCommand>) {
        for arg in &command.groups {
            visitor.visit_group(Arc::clone(&arg));
        }
    }

    pub fn walk_text<T: LatexVisitor>(_visitor: &mut T, _text: Arc<LatexText>) {}

    pub fn walk_comma<T: LatexVisitor>(_visitor: &mut T, _comma: Arc<LatexComma>) {}

    pub fn walk_math<T: LatexVisitor>(_visitor: &mut T, _math: Arc<LatexMath>) {}
}
