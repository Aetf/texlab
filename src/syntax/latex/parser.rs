use crate::syntax::latex::ast::*;
use std::iter::Peekable;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum LatexScope {
    Root,
    Group,
    Options,
    Math,
}

pub struct LatexParser<I: Iterator<Item = LatexToken>> {
    tokens: Peekable<I>,
}

impl<I: Iterator<Item = LatexToken>> LatexParser<I> {
    pub fn new(tokens: I) -> Self {
        LatexParser {
            tokens: tokens.peekable(),
        }
    }

    pub fn root(&mut self) -> LatexRoot {
        let children = self.content(LatexScope::Root);
        LatexRoot::new(children)
    }

    fn content(&mut self, scope: LatexScope) -> Vec<LatexContent> {
        let mut children = Vec::new();
        while let Some(ref token) = self.tokens.peek() {
            match token.kind {
                LatexTokenKind::Word | LatexTokenKind::BeginOptions => {
                    let text = Box::new(self.text(scope));
                    children.push(LatexContent::Text(text));
                }
                LatexTokenKind::Command => {
                    let command = Box::new(self.command());
                    children.push(LatexContent::Command(command));
                }
                LatexTokenKind::Math => {
                    if scope == LatexScope::Math {
                        return children;
                    } else {
                        let group = Box::new(self.group(LatexGroupKind::Math));
                        children.push(LatexContent::Group(group));
                    }
                }
                LatexTokenKind::BeginGroup => {
                    let group = Box::new(self.group(LatexGroupKind::Group));
                    children.push(LatexContent::Group(group));
                }
                LatexTokenKind::EndGroup => {
                    if scope == LatexScope::Root {
                        self.tokens.next();
                    } else {
                        return children;
                    }
                }
                LatexTokenKind::EndOptions => {
                    if scope == LatexScope::Options {
                        return children;
                    } else {
                        let text = Box::new(self.text(scope));
                        children.push(LatexContent::Text(text));
                    }
                }
            }
        }
        children
    }

    fn command(&mut self) -> LatexCommand {
        let name = self.tokens.next().unwrap();
        let options = if self.next_of_kind(LatexTokenKind::BeginOptions) {
            Some(self.group(LatexGroupKind::Options))
        } else {
            None
        };

        let mut args = Vec::new();
        while self.next_of_kind(LatexTokenKind::BeginGroup) {
            args.push(self.group(LatexGroupKind::Group));
        }

        LatexCommand::new(name, options, args)
    }

    fn group(&mut self, kind: LatexGroupKind) -> LatexGroup {
        let left = self.tokens.next().unwrap();
        let scope = match kind {
            LatexGroupKind::Group => LatexScope::Group,
            LatexGroupKind::Options => LatexScope::Options,
            LatexGroupKind::Math => LatexScope::Math,
        };
        let children = self.content(scope);
        let right_kind = match kind {
            LatexGroupKind::Group => LatexTokenKind::EndGroup,
            LatexGroupKind::Options => LatexTokenKind::EndOptions,
            LatexGroupKind::Math => LatexTokenKind::Math,
        };

        let right = if self.next_of_kind(right_kind) {
            self.tokens.next()
        } else {
            None
        };

        LatexGroup::new(left, children, right, kind)
    }

    fn text(&mut self, scope: LatexScope) -> LatexText {
        let mut words = Vec::new();
        while let Some(ref token) = self.tokens.peek() {
            let kind = token.kind;
            let opts = kind == LatexTokenKind::EndOptions && scope != LatexScope::Options;
            if kind == LatexTokenKind::Word || kind == LatexTokenKind::BeginOptions || opts {
                words.push(self.tokens.next().unwrap());
            } else {
                break;
            }
        }
        LatexText::new(words)
    }

    fn next_of_kind(&mut self, kind: LatexTokenKind) -> bool {
        if let Some(ref token) = self.tokens.peek() {
            token.kind == kind
        } else {
            false
        }
    }
}
