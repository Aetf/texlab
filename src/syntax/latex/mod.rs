mod ast;
mod finder;
mod lexer;
mod parser;

pub use self::ast::*;
use self::finder::LatexFinder;
pub use self::finder::LatexNode;
use crate::data::language::*;
use crate::syntax::latex::lexer::LatexLexer;
use crate::syntax::latex::parser::LatexParser;
use crate::syntax::text::SyntaxNode;
use lsp_types::{Position, Range, Uri};
use path_clean::PathClean;
use std::path::PathBuf;
use std::sync::Arc;

#[derive(Debug, Default)]
struct LatexCommandAnalyzer {
    commands: Vec<Arc<LatexCommand>>,
}

impl LatexCommandAnalyzer {
    fn parse(root: Arc<LatexRoot>) -> Vec<Arc<LatexCommand>> {
        let mut analyzer = Self::default();
        analyzer.visit_root(root);
        analyzer.commands
    }
}

impl LatexVisitor for LatexCommandAnalyzer {
    fn visit_root(&mut self, root: Arc<LatexRoot>) {
        LatexWalker::walk_root(self, root);
    }

    fn visit_group(&mut self, group: Arc<LatexGroup>) {
        LatexWalker::walk_group(self, group);
    }

    fn visit_command(&mut self, command: Arc<LatexCommand>) {
        self.commands.push(Arc::clone(&command));
        LatexWalker::walk_command(self, command);
    }

    fn visit_text(&mut self, text: Arc<LatexText>) {
        LatexWalker::walk_text(self, text);
    }

    fn visit_comma(&mut self, comma: Arc<LatexComma>) {
        LatexWalker::walk_comma(self, comma);
    }

    fn visit_math(&mut self, math: Arc<LatexMath>) {
        LatexWalker::walk_math(self, math);
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexEnvironmentDelimiter {
    pub command: Arc<LatexCommand>,
}

impl LatexEnvironmentDelimiter {
    pub fn name(&self) -> Option<&LatexToken> {
        self.command.extract_word(0)
    }
}

impl SyntaxNode for LatexEnvironmentDelimiter {
    fn range(&self) -> Range {
        self.command.range()
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexEnvironment {
    pub left: LatexEnvironmentDelimiter,
    pub right: LatexEnvironmentDelimiter,
}

impl LatexEnvironment {
    pub fn is_root(&self) -> bool {
        self.left
            .name()
            .iter()
            .chain(self.right.name().iter())
            .any(|name| name.text() == "document")
    }

    fn parse(commands: &[Arc<LatexCommand>]) -> Vec<Self> {
        let mut stack = Vec::new();
        let mut environments = Vec::new();
        for command in commands {
            if let Some(delimiter) = Self::parse_delimiter(command) {
                if delimiter.command.name.text() == "\\begin" {
                    stack.push(delimiter);
                } else if let Some(begin) = stack.pop() {
                    environments.push(Self {
                        left: begin,
                        right: delimiter,
                    });
                }
            }
        }
        environments
    }

    fn parse_delimiter(command: &Arc<LatexCommand>) -> Option<LatexEnvironmentDelimiter> {
        if command.name.text() != "\\begin" && command.name.text() != "\\end" {
            return None;
        }

        if command.args.len() == 0 {
            return None;
        }

        if command.has_word(0) || command.args[0].children.is_empty() {
            let delimiter = LatexEnvironmentDelimiter {
                command: Arc::clone(&command),
            };
            Some(delimiter)
        } else {
            None
        }
    }
}

impl SyntaxNode for LatexEnvironment {
    fn range(&self) -> Range {
        Range::new(self.left.start(), self.right.end())
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexCitation {
    pub command: Arc<LatexCommand>,
    pub index: usize,
}

impl LatexCitation {
    pub fn keys(&self) -> Vec<&LatexToken> {
        self.command.extract_comma_separated_words(0)
    }

    fn parse(commands: &[Arc<LatexCommand>]) -> Vec<Self> {
        let mut citations = Vec::new();
        for command in commands {
            for LatexCitationCommand { name, index } in &language_data().citation_commands {
                if command.name.text() == name && command.has_comma_separated_words(*index) {
                    citations.push(Self {
                        command: Arc::clone(command),
                        index: *index,
                    });
                }
            }
        }
        citations
    }
}

impl SyntaxNode for LatexCitation {
    fn range(&self) -> Range {
        self.command.range()
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexLabel {
    pub command: Arc<LatexCommand>,
    index: usize,
    pub kind: LatexLabelKind,
}

impl LatexLabel {
    pub fn names(&self) -> Vec<&LatexToken> {
        self.command.extract_comma_separated_words(self.index)
    }

    fn parse(commands: &[Arc<LatexCommand>]) -> Vec<Self> {
        let mut labels = Vec::new();
        for command in commands {
            for LatexLabelCommand { name, index, kind } in &language_data().label_commands {
                if command.name.text() == name && command.has_comma_separated_words(*index) {
                    labels.push(Self {
                        command: Arc::clone(command),
                        index: *index,
                        kind: *kind,
                    });
                }
            }
        }
        labels
    }
}

impl SyntaxNode for LatexLabel {
    fn range(&self) -> Range {
        self.command.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexSection {
    pub command: Arc<LatexCommand>,
    pub index: usize,
    pub level: i32,
}

impl LatexSection {
    fn parse(commands: &[Arc<LatexCommand>]) -> Vec<Self> {
        let mut sections = Vec::new();
        for command in commands {
            for LatexSectionCommand { name, index, level } in &language_data().section_commands {
                if command.name.text() == name && command.args.len() > *index {
                    sections.push(Self {
                        command: Arc::clone(command),
                        index: *index,
                        level: *level,
                    })
                }
            }
        }
        sections
    }
}

impl SyntaxNode for LatexSection {
    fn range(&self) -> Range {
        self.command.range()
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexInclude {
    pub command: Arc<LatexCommand>,
    pub index: usize,
    pub kind: LatexIncludeKind,
    pub targets: Vec<Uri>,
    pub include_extension: bool,
}

impl LatexInclude {
    pub fn path(&self) -> &LatexToken {
        self.command.extract_word(self.index).unwrap()
    }

    pub fn component_name(&self) -> Option<String> {
        match self.kind {
            LatexIncludeKind::Package => Some(format!("{}.sty", self.path().text())),
            LatexIncludeKind::Class => Some(format!("{}.cls", self.path().text())),
            LatexIncludeKind::Latex
            | LatexIncludeKind::Bibliography
            | LatexIncludeKind::Image
            | LatexIncludeKind::Svg
            | LatexIncludeKind::Everything => None,
        }
    }

    fn parse(uri: &Uri, commands: &[Arc<LatexCommand>]) -> Vec<Self> {
        let mut includes = Vec::new();
        for command in commands {
            for description in &language_data().include_commands {
                if let Some(include) = Self::parse_single(uri, &command, &description) {
                    includes.push(include);
                }
            }
        }
        includes
    }

    fn parse_single(
        uri: &Uri,
        command: &Arc<LatexCommand>,
        description: &LatexIncludeCommand,
    ) -> Option<Self> {
        if command.name.text() != description.name {
            return None;
        }

        let relative_path = command.extract_word(0)?;
        let mut path = uri.to_file_path().ok()?;
        path.pop();
        path.push(relative_path.text());
        path = PathBuf::from(path.to_string_lossy().into_owned().replace('\\', "/"));
        path = path.clean();
        let mut path = path.to_str()?.to_owned();

        let mut targets = Vec::new();
        targets.push(Uri::from_file_path(&path).ok()?);
        if let Some(extensions) = description.kind.extensions() {
            for extension in extensions {
                let path = format!("{}.{}", &path, extension);
                targets.push(Uri::from_file_path(&path).ok()?);
            }
        }

        let include = Self {
            command: Arc::clone(command),
            index: description.index,
            kind: description.kind,
            targets,
            include_extension: description.include_extension,
        };
        Some(include)
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexEquation {
    pub left: Arc<LatexCommand>,
    pub right: Arc<LatexCommand>,
}

impl LatexEquation {
    fn parse(commands: &[Arc<LatexCommand>]) -> Vec<Self> {
        let mut equations = Vec::new();
        let mut left = None;
        for command in commands {
            let name = command.name.text();
            if name == "\\[" || name == "\\(" {
                left = Some(command);
            } else if name == "\\]" || name == "\\)" {
                if let Some(begin) = left {
                    equations.push(Self {
                        left: Arc::clone(&begin),
                        right: Arc::clone(&command),
                    });
                    left = None;
                }
            }
        }
        equations
    }
}

impl SyntaxNode for LatexEquation {
    fn range(&self) -> Range {
        Range::new(self.left.start(), self.right.end())
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexInline {
    pub left: Arc<LatexMath>,
    pub right: Arc<LatexMath>,
}

impl LatexInline {
    fn parse(root: Arc<LatexRoot>) -> Vec<Self> {
        let mut analyzer = LatexInlineAnalyzer::default();
        analyzer.visit_root(root);
        analyzer.inlines
    }
}

impl SyntaxNode for LatexInline {
    fn range(&self) -> Range {
        Range::new(self.left.start(), self.right.end())
    }
}

#[derive(Debug, Default)]
struct LatexInlineAnalyzer {
    inlines: Vec<LatexInline>,
    left: Option<Arc<LatexMath>>,
}

impl LatexVisitor for LatexInlineAnalyzer {
    fn visit_root(&mut self, root: Arc<LatexRoot>) {
        LatexWalker::walk_root(self, root);
    }

    fn visit_group(&mut self, group: Arc<LatexGroup>) {
        LatexWalker::walk_group(self, group);
    }

    fn visit_command(&mut self, command: Arc<LatexCommand>) {
        LatexWalker::walk_command(self, command);
    }

    fn visit_text(&mut self, text: Arc<LatexText>) {
        LatexWalker::walk_text(self, text);
    }

    fn visit_comma(&mut self, comma: Arc<LatexComma>) {
        LatexWalker::walk_comma(self, comma);
    }

    fn visit_math(&mut self, math: Arc<LatexMath>) {
        if let Some(left) = &self.left {
            let inline = LatexInline {
                left: Arc::clone(&left),
                right: Arc::clone(&math),
            };
            self.inlines.push(inline);
            self.left = None;
        } else {
            self.left = Some(Arc::clone(&math));
        }
        LatexWalker::walk_math(self, math);
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexMathOperator {
    pub command: Arc<LatexCommand>,
    pub definition_index: usize,
    pub implementation_index: usize,
}

impl LatexMathOperator {
    fn parse(commands: &[Arc<LatexCommand>]) -> Vec<Self> {
        let mut operators = Vec::new();
        for command in commands {
            for LatexMathOperatorCommand {
                name,
                definition_index,
                implementation_index,
            } in &language_data().math_operator_commands
            {
                if command.name.text() == name
                    && command.args.len() > *definition_index
                    && command.args.len() > *implementation_index
                {
                    operators.push(Self {
                        command: Arc::clone(command),
                        definition_index: *definition_index,
                        implementation_index: *implementation_index,
                    })
                }
            }
        }
        operators
    }
}

impl SyntaxNode for LatexMathOperator {
    fn range(&self) -> Range {
        self.command.range()
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexCommandDefinition {
    pub command: Arc<LatexCommand>,
    pub definition: Arc<LatexCommand>,
    pub definition_index: usize,
    pub implementation: Arc<LatexGroup>,
    pub implementation_index: usize,
    pub argument_count_index: usize,
}

impl LatexCommandDefinition {
    fn parse(commands: &[Arc<LatexCommand>]) -> Vec<Self> {
        let mut definitions = Vec::new();
        for command in commands {
            for LatexCommandDefinitionCommand {
                name,
                definition_index,
                argument_count_index,
                implementation_index,
            } in &language_data().command_definition_commands
            {
                if command.name.text() == name
                    && command.args.len() > *definition_index
                    && command.args.len() > *implementation_index
                {
                    let name = command.args[0].children.iter().next();
                    if let Some(LatexContent::Command(name)) = name {
                        definitions.push(Self {
                            command: Arc::clone(command),
                            definition: Arc::clone(name),
                            definition_index: *definition_index,
                            implementation: Arc::clone(&command.args[*implementation_index]),
                            implementation_index: *implementation_index,
                            argument_count_index: *argument_count_index,
                        })
                    }
                }
            }
        }
        definitions
    }
}

impl SyntaxNode for LatexCommandDefinition {
    fn range(&self) -> Range {
        self.command.range()
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexSyntaxTree {
    pub root: Arc<LatexRoot>,
    pub commands: Vec<Arc<LatexCommand>>,
    pub includes: Vec<LatexInclude>,
    pub components: Vec<String>,
    pub environments: Vec<LatexEnvironment>,
    pub is_standalone: bool,
    pub labels: Vec<LatexLabel>,
    pub sections: Vec<LatexSection>,
    pub citations: Vec<LatexCitation>,
    pub equations: Vec<LatexEquation>,
    pub inlines: Vec<LatexInline>,
    pub math_operators: Vec<LatexMathOperator>,
    pub command_definitions: Vec<LatexCommandDefinition>,
}

impl LatexSyntaxTree {
    pub fn parse(uri: &Uri, text: &str) -> Self {
        let lexer = LatexLexer::new(text);
        let mut parser = LatexParser::new(lexer);
        let root = Arc::new(parser.root());
        let commands = LatexCommandAnalyzer::parse(Arc::clone(&root));
        let includes = LatexInclude::parse(uri, &commands);
        let components = includes
            .iter()
            .flat_map(LatexInclude::component_name)
            .collect();
        let environments = LatexEnvironment::parse(&commands);
        let is_standalone = environments.iter().any(LatexEnvironment::is_root);
        let labels = LatexLabel::parse(&commands);
        let sections = LatexSection::parse(&commands);
        let citations = LatexCitation::parse(&commands);
        let equations = LatexEquation::parse(&commands);
        let inlines = LatexInline::parse(Arc::clone(&root));
        let math_operators = LatexMathOperator::parse(&commands);
        let command_definitions = LatexCommandDefinition::parse(&commands);
        Self {
            root,
            commands,
            includes,
            components,
            environments,
            is_standalone,
            labels,
            sections,
            citations,
            equations,
            inlines,
            math_operators,
            command_definitions,
        }
    }

    pub fn find(&self, position: Position) -> Vec<LatexNode> {
        let mut finder = LatexFinder::new(position);
        finder.visit_root(Arc::clone(&self.root));
        finder.results
    }

    pub fn find_command(&self, position: Position) -> Option<Arc<LatexCommand>> {
        for result in self.find(position) {
            if let LatexNode::Command(command) = result {
                if command.name.range().contains(position)
                    && command.name.start().character != position.character
                {
                    return Some(command);
                }
            }
        }
        None
    }
}
