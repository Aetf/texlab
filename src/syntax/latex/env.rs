use super::ast::*;
use crate::syntax::language::*;
use crate::syntax::text::SyntaxNode;
use lsp_types::Range;
use std::sync::Arc;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexEnvironmentDelimiter {
    pub command: Arc<LatexCommand>,
}

impl LatexEnvironmentDelimiter {
    pub fn name(&self) -> Option<&LatexToken> {
        self.command.extract_word(0)
    }

    pub fn is_math(&self) -> bool {
        if let Some(name) = self.name() {
            LANGUAGE_DATA
                .math_environments
                .iter()
                .any(|env| env == name.text())
        } else {
            false
        }
    }

    pub fn is_enum(&self) -> bool {
        if let Some(name) = self.name() {
            LANGUAGE_DATA
                .enum_environments
                .iter()
                .any(|env| env == name.text())
        } else {
            false
        }
    }
}

impl SyntaxNode for LatexEnvironmentDelimiter {
    fn range(&self) -> Range {
        self.command.range()
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexEnvironment {
    pub left: LatexEnvironmentDelimiter,
    pub right: LatexEnvironmentDelimiter,
}

impl LatexEnvironment {
    pub fn is_root(&self) -> bool {
        self.left
            .name()
            .iter()
            .chain(self.right.name().iter())
            .any(|name| name.text() == "document")
    }

    fn parse(commands: &[Arc<LatexCommand>]) -> Vec<Self> {
        let mut stack = Vec::new();
        let mut environments = Vec::new();
        for command in commands {
            if let Some(delimiter) = Self::parse_delimiter(command) {
                if delimiter.command.name.text() == "\\begin" {
                    stack.push(delimiter);
                } else if let Some(begin) = stack.pop() {
                    environments.push(Self {
                        left: begin,
                        right: delimiter,
                    });
                }
            }
        }
        environments
    }

    fn parse_delimiter(command: &Arc<LatexCommand>) -> Option<LatexEnvironmentDelimiter> {
        if command.name.text() != "\\begin" && command.name.text() != "\\end" {
            return None;
        }

        if command.args.is_empty() {
            return None;
        }

        if command.has_word(0)
            || command.args[0].children.is_empty()
            || command.args[0].right.is_none()
        {
            let delimiter = LatexEnvironmentDelimiter {
                command: Arc::clone(&command),
            };
            Some(delimiter)
        } else {
            None
        }
    }
}

impl SyntaxNode for LatexEnvironment {
    fn range(&self) -> Range {
        Range::new(self.left.start(), self.right.end())
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexEnvironmentInfo {
    pub environments: Vec<LatexEnvironment>,
    pub is_standalone: bool,
}

impl LatexEnvironmentInfo {
    pub fn parse(commands: &[Arc<LatexCommand>]) -> Self {
        let environments = LatexEnvironment::parse(commands);
        let is_standalone = environments.iter().any(LatexEnvironment::is_root);
        Self {
            environments,
            is_standalone,
        }
    }
}
