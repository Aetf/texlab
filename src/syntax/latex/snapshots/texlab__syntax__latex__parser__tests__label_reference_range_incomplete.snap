---
source: src/syntax/latex/parser.rs
expression: "setup(r#\"\\crefrange{foo}\"#)"

---
ROOT@0..15
  PREAMBLE@0..15
    LABEL_REFERENCE_RANGE@0..15
      LABEL_REFERENCE_RANGE_NAME@0..10 "\\crefrange"
      CURLY_GROUP_WORD@10..15
        L_CURLY@10..11 "{"
        KEY@11..14
          WORD@11..14 "foo"
        R_CURLY@14..15 "}"
      MISSING@15..15 ""

